swarm/src/handler.rs

```diff
@@ -508,25 +496,53 @@ impl<TConnectionUpgrade, TOutboundOpenInfo, TCustom, TErr>
             ConnectionHandlerEvent::Close(val) => ConnectionHandlerEvent::Close(val),
         }
     }
+}
 
-    /// If this is a `Close` event, maps the content to something else.
-    pub fn map_close<F, I>(
-        self,
-        map: F,
-    ) -> ConnectionHandlerEvent<TConnectionUpgrade, TOutboundOpenInfo, TCustom, I>
-    where
-        F: FnOnce(TErr) -> I,
-    {
-        match self {
-            ConnectionHandlerEvent::OutboundSubstreamRequest { protocol } => {
-                ConnectionHandlerEvent::OutboundSubstreamRequest { protocol }
-            }
-            ConnectionHandlerEvent::Custom(val) => ConnectionHandlerEvent::Custom(val),
-            ConnectionHandlerEvent::Close(val) => ConnectionHandlerEvent::Close(map(val)),
+#[derive(Debug)]
+pub struct CloseReason {
+    component_name: &'static str,
+    source: Box<dyn error::Error + Send + 'static>,
+}
+
+impl CloseReason {
+    /// Construct a new [`CloseReason`].
+    ///
+    /// The first parameter should be a meaningful identifier for the component / protocol that is closing the connection.
+    /// The given `source` is returned from [`Error::source`](error::Error::source) and can be printed by iterating the source of this error.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use libp2p_swarm::handler::CloseReason;
+    ///
+    /// # fn main() {
+    /// let source = std::io::Error::from(std::io::ErrorKind::UnexpectedEof); // Imagine this being the result of reading from a stream.
+    /// let reason = CloseReason::new("ping", source);
+    ///
+    /// assert_eq!(reason.to_string(), "connection closed by 'ping' protocol")
+    /// # }
+    ///
+    /// ```
+    pub fn new(component_name: &'static str, source: impl error::Error + Send + 'static) -> Self {
+        Self {
+            component_name,
+            source: Box::new(source),
         }
     }
 }
 
+impl fmt::Display for CloseReason {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "connection closed by '{}' protocol", self.component_name)
+    }
+}
+
+impl error::Error for CloseReason {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        Some(self.source.as_ref())
```

Comment:
Perhaps document that logging these errors should follow the `.source()` chain, e.g. by using `anyhow` and formatting with `{:#}`. Or is there a generic helper for implementing alternate Error printing so we can offer it out of the box? (This is the weakest point in Rust for me, probably since I’m spoilt by living on the JVM for so long.)

Reply:
I don't think there is unfortunately but I whole-heartedly agree that it is really missing from std.

We could offer it on this type-specifically although I think it would just be better to tell users to use anyhow.

Reply:
I don't think there is unfortunately but I whole-heartedly agree that it is really missing from std.

We could offer it on this type-specifically although I think it would just be better to tell users to use anyhow.

Reply:
Agreed, let’s document it.